Modules/StudyProgramme/classes/model/Progress/class.ilStudyProgrammeProgress.php

```diff
@@ -287,203 +289,315 @@ public function getStatus() : int
-     * @param DateTime | null	$deadline
-     *
-     * @return $this
-     */
-    public function setDeadline(DateTime $deadline = null) : ilStudyProgrammeProgress
+    public function hasIndividualModifications() : bool
     {
-        $this->deadline = $deadline;
-        return $this;
+        return $this->is_individual;
     }
 
-    /**
-     * Set limited validity of qualification date.
-     */
-    public function setValidityOfQualification(DateTime $date = null) : ilStudyProgrammeProgress
+    public function withIndividualModifications(bool $individual) : ilStudyProgrammeProgress
     {
-        $this->vq_date = $date;
-        return $this;
+        $clone = clone $this;
+        $clone->is_individual = $individual;
+        return $clone;
     }
 
+    public function isSuccessful() : bool
+    {
+        return in_array(
+            $this->getStatus(),
+            [
+                self::STATUS_COMPLETED,
+                self::STATUS_ACCREDITED
+            ]
+        );
+    }
+    
     /**
-     * Get the limited validity of qualification date.
+     * There may be no qualification at all (since the PRG is not passed),
+     * or the qualification is valid or invalid due to a date.
      */
-    public function getValidityOfQualification() : ?DateTime
+    public function hasValidQualification(DateTimeImmutable $now) : ?bool
     {
-        return $this->vq_date;
+        if (!$this->isSuccessful()) {
+            return null;
+        }
+        return (
+            is_null($this->getValidityOfQualification()) ||
+            $this->getValidityOfQualification()->format('Y-m-d') >= $now->format('Y-m-d')
+        );
     }
 
-    /**
-     * Toggle invalidated.
-     */
+    public function isRelevant() : bool
+    {
+        return $this->getStatus() != self::STATUS_NOT_RELEVANT;
+    }
+
+    public function isFailed() : bool
+    {
+        return $this->getStatus() == self::STATUS_FAILED;
+    }
+    
+    public function isAccredited() : bool
+    {
+        return $this->getStatus() == self::STATUS_ACCREDITED;
+    }
+
+    public function isInProgress() : bool
+    {
+        return $this->getStatus() == self::STATUS_IN_PROGRESS;
+    }
+    
     public function invalidate() : ilStudyProgrammeProgress
     {
         if (!$this->vq_date || $this->vq_date->format('Y-m-d') > date('Y-m-d')) {
             throw new ilException("may not invalidate non-expired progress");
         }
-        $this->invalidated = true;
-        return $this;
+        $clone = clone $this;
+        $clone->invalidated = true;
+        return $clone;
     }
-
+    
     public function isInvalidated() : bool
     {
         return $this->invalidated;
     }
+
+    /**
+     * @deprecated
+     */
+    public function isSuccessfulExpired() : bool
+    {
+        if (
+            !is_null($this->getValidityOfQualification()) &&
+            $this->getValidityOfQualification()->format('Y-m-d') < (new DateTimeImmutable())->format('Y-m-d')
```

Comment:
This will break assumptions about the value-object, because a `new` is involved here, which will silently change the internal state of this object. Either parametrize the method with that datetime or move this check elsewhere.